lib/pocolog/stream_index.rb

```diff
@@ -47,15 +48,46 @@ def base_time=(value)
             offset = @base_time - value
             return if offset == 0
 
-            @index_map = index_map.map do |file_pos, time, sample_index|
-                [file_pos, time + offset, sample_index]
-            end
+            @index_map = map_entries_internal { |pos, t| [pos, t + offset] }
             @base_time = value
         end
 
+        # Create a new index object with its entries changed
+        #
+        # @yieldparam [Integer] pos the file position for this sample
+        # @yieldparam [Integer] time the internal sample time, as an offset in
+        #   microseconds from {#base_time}
+        def map_entries(&block)
+            new_map = map_entries_internal(&block)
+            self.class.from_raw_data(@base_time, new_map)
+        end
+
+        # @api private
+        #
+        # Non-destructively change the internal @index_map object
+        def map_entries_internal(&block)
+            self.class.map_entries_internal(@index_map, &block)
+        end
+
+        # @api private
+        #
+        # Map entries of an index index map data array
+        #
+        # @yieldparam [Integer] pos file position
+        # @yieldparam [Integer] time internal time
+        # @yieldreturn [(Integer, Integer)] new position and time
+        # @return [Array] updated index map
+        def self.map_entries_internal(index_map)
+            new_map = Array.new(index_map.size)
+            (0...index_map.size).step(2) do |i|
+                new_map[i, 2] = yield(index_map[i, 2])
+            end
+            new_map
```

Comment:
Ins't `index_map` flattened?

Reply:
It is ... new_map[i, 2] sets the two elements at i and i+1